types/address.go

```diff
@@ -256,6 +264,31 @@ func (aa AccAddress) Format(s fmt.State, verb rune) {
 	}
 }
 
+// LengthPrefixedAddress prefixes the address bytes with its length, this is used
+// for variable-length components in store keys. Note: All addresses should be
+// max 255 bytes, or else this function panics.
+func LengthPrefixedAddress(bz []byte) ([]byte, error) {
+	if len(bz) == 0 {
+		return bz, nil
+	}
+
+	if len(bz) > maxAddrLen {
+		return nil, sdkerrors.Wrapf(sdkerrors.ErrUnknownAddress, "address length should be max %d bytes, got %d", maxAddrLen, len(bz))
+	}
+
+	return append([]byte{byte(len(bz))}, bz...), nil
+}
```

Comment:
I think we need to add 2 more checks:
1. `lenBz` can't be 19 -> this would create an address which is 20 bytes and will conflict with legacy addresses
2. `lenBz` can't be 31 -> this would create an address which is 32 bytes and will conflict with new hash addresses

Not sure though, if we will do implement any heuristics for checking the length and decoding a right address.

Reply:
These conflicts are not an issue here. This is for store keys. This function just should be renamed

Reply:
Let's maybe call this LengthPrefixedAddressStoreKey or something @amaurymartiny ?

Reply:
Related to https://github.com/cosmos/cosmos-sdk/pull/8363#discussion_r565232303 : At the end of day, this function doesn't care if the bz are an Address or not. So maybe `LengthPrefixedStoreKeyComponent`/`LengthPrefixedStoreKeyPart`?

Most importantly, i'm not sure it should go in `types/address/keys.go` (like Robert suggested) or `types/address.go`, I would just put it `types/keys.go`.

Reply:
`/types` package is bloated. It not only contains _some_ types, but also functions. I would prefer to avoid adding things to `/types` whenever possible.

Reply:
IMHO, `LengthPrefixedAddressStoreKey` is too long. If we put it in the `address` package, then we will have a context and we can make it much shorter: `address.MkLenthPrefixed`, `address.PrefixLength` or `address.WithLength`. The last one is my preferred.

Reply:
What about `address.NewWithLength()`?

Reply:
I renamed to `address.LengthPrefixedStoreKey`, I hope that's okay for everyone.

Most importantly, I wanted the name to show that it's should solely be used in store keys, and not in any other places.

Reply:
It's fine. Personally I like short names whenever possible.